### ISSUE TITLE
bug: "Could not connect to robot" dialog appears and disappears at confusing times

### ISSUE THREAD

**Issue by SyntaxColoring** [Opentrons/opentrons#5240]
# Overview

* The "Could not connect to robot" dialog sometimes disappears before the user has a chance to read it.
* The "Could not connect to robot" dialog sometimes appears when the user *successfully gains* a connection to a robot.

# Details of current behavior

I think these behaviors are the result of an unintended interaction between two error handling systems.

When a robot is in the list and you try to connect to it, but the network connection is broken, two things will happen:

1. The connection request will eventually time out. When that happens, the app will show a "Could not connect to robot" modal dialog.
2. Separately, the Opentrons App will eventually notice that it hasn't heard anything from the robot in a while. When that happens, it will remove the robot from the list.

These two things race against each other.

Furthermore, the dialog behaves as a part of that specific robot's page, rather than a part of the entire app window. Meaning, when the robot's page is hidden, the dialog is hidden along with it; and, when the robot's page comes back, the dialog will come back too.

Combined, this can cause the following confusing sequence:

1. You try to connect to a robot over a flaky connection.
2. **You may or may not see a "Could not connect to robot" dialog.**
3. The robot is removed from the list, and you're kicked out of its page. **If you saw a "Could not connect to robot," it's wiped away.**
4. The connection is fixed. (The microwave in the next room turns off, the USB cable is plugged back in, whatever.)
5. **You reconnect to the robot and immediately see a "Could not connect to robot" dialog.** The dialog is stale, referring to when you lost connection before. But the way it appears makes it look like it's referring to the connection that you just regained.

Although I haven't tested it, I assume this also happens for other interactions that take a network request, like turning on the deck lights or homing the robot.

# Steps to reproduce

1. Have a robot plugged in over USB and ready to connect to in the Opentrons App.
2. Unplug the USB cable from your computer.
3. Immediately try to connect to that robot.
4. Wait. A "Could not connect to robot" dialog may appear. Leave it up.
5. Wait some more. The robot will be removed from the list because the app hasn't heard from it in a while. Simultaneously, the "Could not connect" dialog will go away.
6. Plug the USB cable back in.
7. Click "Try again" once or twice as needed to regain the connection.
8. When the robot reappears in the list, its page will auto-open, and the "Could not connect" dialog will reappear.

Tested on v3.16.1, disconnected from Wi-Fi. (So the one robot is 

# Expected behavior

Either:
* The "Could not connect" dialog should persist until the user dismisses it. This solves both the flash-of-dialog problem and the stale-dialog problem.
* Or, the "Could not connect" dialog should be cleared along with the robot page, so that it doesn't reappear when you reconnect. This solves only the "stale-dialog" problem, but I consider that to be the more important of the two.
* Or, we use something other than a modal dialog to report a connection failure. It seems like modals are tricky to get right for this kind of error. See also issue #5241.

**Comment by mcous** [comment]
## TL;DR: solution notes

- Do not re-direct route if robot is not connectable; keep robot name in route and display splash in robot page component
    - This should ensure any required state (the robot name) for displaying alerts remains in place, so modals don't disappear along with the contents of the page

Implementation / fix notes:

> 1. You try to connect to a robot over a flaky connection.

User has clicked the "connect" button or toggle, which will cause the app to attempt to open a WebSocket for the RPC server

> 2. You may or may not see a "Could not connect to robot" dialog.

User will see this if the WebSocket fails to open or if the RPC server fails to negotiate the initiate the connection properly

> 3. The robot is removed from the list, and you're kicked out of its page. If you saw a "Could not connect to robot," it's wiped away.

As noted, this is independent of the WebSocket stuff. We don't have UI for a robot page for a robot that is not connectable. I think this would be a good thing to have for a variety of reasons, so I'll likely open another ticket and drop it here

> 4. The connection is fixed. (The microwave in the next room turns off, the USB cable is plugged back in, whatever.)

**Important to note**: referring here to the network connection, not the specific RPC WebSocket

> 5. You reconnect to the robot and immediately see a "Could not connect to robot" dialog. The dialog is stale, referring to when you lost connection before. But the way it appears makes it look like it's referring to the connection that you just regained.

**Also important to note:** When this happens, the user **has not reconnected the RPC WebSocket**. The **networking** connection has become good enough again to allow HTTP `GET /health` to succeed, so the robot page is displayed again. Certainly the user thinks they have "reconnected", but at this point the app has not established a persistent WebSocket connection with the robot.

So, in a technical sense, the "could not connect to robot" dialog is not stale here. The last time the user tried to establish an RPC WebSocket connection to this robot, it failed, and the user never explicitly dismissed that dialog, so it displays.

> Although I haven't tested it, I assume this also happens for other interactions that take a network request, like turning on the deck lights or homing the robot.

These interactions are different, because they happen over HTTP requests and do not involve the persistent RPC WebSocket. However, if the HTTP request goes out and fails, that error message will still be present if the robot drops off the list and comes back.

**Comment by mcous** [comment]
As of 3.20.0, the app will not remove any robot from the UI that it's been talking to while the app has been open. Consequently, the "Could not connect" modal will not disappear until the user clicks the close button
